### PR TITLE
Add support for FGAT and IAU for aerosol DA

### LINKED ISSUE #1930: Add support for FGAT for aerosol DA
### What new functionality do you need?

The FV3 model in the gdas cycle already writes out RESTART files at fh003, fh006, and fh009. I believe the workflow saves all of these (definitely the first two) in the rotdir. The aerosol job tasks need modified to copy/link the fh003 and fh009 files (along with GDASApp related changes) to allow for 3DFGAT for the JEDI-based 3DVar aerosol analysis. This is related to https://github.com/NOAA-EMC/GDASApp/issues/568

### What are the requirements for the new functionality?

FMS RESTART files from GDAS forecast are available in the ROTDIR; the aeroanlinit job links/copies multiple RESTART files

### Acceptance Criteria

when we can exercise 3DFGAT in the aerosol DA cycling system

### Suggest a solution (optional)

_No response_

### PR DESCRIPTION
**Pull request by CoryMartin-NOAA**
# Description
This PR adds support for FGAT for the 3DVar aerosol DA. Not only does this allow for FGAT but it also allows for atmospheric IAU to run with aerosol DA.

Resolves #1930

# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? No
- Does this change require a documentation update? Maybe?

# How has this been tested?
- Cycled test on Hera

# Checklist
- [ ] Any dependent changes have been merged and published
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

**Comment by CoryMartin-NOAA** [comment]
This is draft until I can add the commit with the updated GDASApp hash

File: parm/config/gfs/config.aeroanl [hunk 1]
```diff
@@ -20,4 +19,12 @@ export io_layout_y=@IO_LAYOUT_Y@
 
 export JEDIEXE=${HOMEgfs}/exec/fv3jedi_var.x
 
+if [[ "${DOIAU}" == "YES" ]]; then
+  export restart_interval="3"
+  export AEROVARYAML=${HOMEgfs}/sorc/gdas.cd/parm/aero/variational/3dvar_fgat_gfs_aero.yaml
```
Comment (CoryMartin-NOAA) [inline parm/config/gfs/config.aeroanl:24]: I will work on linking these things to the workflow root directory in a subsequent PR.

**Review by andytangborn — APPROVED** [review]
Looks fine.  I take it that nowtime is introduced because the window has several times to use for creating a list of files to read in.

**Comment by CoryMartin-NOAA** [comment]
Besides a GDASApp hash for `feature/stable-nightly` (that hopefully comes in overnight), this is ready for review, as the OOPS PR has just been merged.

File: parm/config/gfs/config.aeroanl [hunk 2]
```diff
@@ -20,4 +19,12 @@ export io_layout_y=@IO_LAYOUT_Y@
 
 export JEDIEXE=${HOMEgfs}/exec/fv3jedi_var.x
 
+if [[ "${DOIAU}" == "YES" ]]; then
+  export restart_interval="3"
```
Comment (aerorahul) [inline parm/config/gfs/config.aeroanl]: where is this variable `restart_interval` used in the aeroanl jobs?

File: ush/python/pygfs/task/aero_analysis.py [hunk 3]
```diff
@@ -167,10 +165,12 @@ def finalize(self: Analysis) -> None:
...
         FileHandler(yaml_copy).sync()
 
-        # ---- NOTE below is 'temporary', eventually we will not be using FMS RESTART formatted files
-        # ---- all of the rest of this method will need to be changed but requires model and JEDI changes
         # ---- copy RESTART fv_tracer files for future reference
-        template = '{}.fv_tracer.res.tile{}.nc'.format(to_fv3time(self.task_config.current_cycle), '{tilenum}')
+        if self.task_config.DOIAU:
+            restarttime = self.task_config.AERO_WINDOW_BEGIN
+        else:
+            restarttime = self.task_config.current_cycle
```
Comment (aerorahul) [inline ush/python/pygfs/task/aero_analysis.py]: Should this be elevated to `__init__` and compute as `model_start_time` once and for all?
Perhaps in `__init__` we can also compute `background_times` list once.
Thoughts?

File: ush/python/pygfs/task/aero_analysis.py [hunk 4]
```diff
@@ -200,13 +200,16 @@ def clean(self):
     @logit(logger)
     def _add_fms_cube_sphere_increments(self: Analysis) -> None:
         """This method adds increments to RESTART files to get an analysis
-        NOTE this is only needed for now because the model cannot read aerosol increments.
-        This method will be assumed to be deprecated before this is implemented operationally
         """
+        if self.task_config.DOIAU:
+            restarttime = self.task_config.AERO_WINDOW_BEGIN
+        else:
+            restarttime = self.task_config.current_cycle
```
Comment (aerorahul) [inline ush/python/pygfs/task/aero_analysis.py]: If this is elevated to `__init__` as in the previous comment, we won't need to keep calculating it.

**Review by aerorahul — COMMENTED** [review]
some minor comments.

File: ush/python/pygfs/task/aero_analysis.py (hunk 3 above)
Comment (CoryMartin-NOAA) [reply ush/python/pygfs/task/aero_analysis.py]: I think that's a good suggestion. Do we want that as part of this PR or separate?

File: ush/python/pygfs/task/aero_analysis.py [hunk 5]
```diff
@@ -238,16 +241,29 @@ def get_bkg_dict(self, task_config: Dict[str, Any]) -> Dict[str, List[str]]:
...
-            for itile in range(1, task_config.ntiles + 1):
-                basename = template.format(tilenum=itile)
-                bkglist.append([os.path.join(rst_dir, basename), os.path.join(run_dir, basename)])
+        # if using IAU, we can use FGAT
+        bkgtimes = []
+        if task_config.DOIAU:
+            nowtime = task_config.AERO_WINDOW_BEGIN
+            while nowtime <= add_to_datetime(task_config.current_cycle, to_timedelta(f"{task_config['assim_freq']}H") / 2):
+                bkgtimes.append(nowtime)
+                nowtime = add_to_datetime(nowtime, to_timedelta(f"{task_config['restart_interval']}H"))
```
Comment (CoryMartin-NOAA) [inline ush/python/pygfs/task/aero_analysis.py]: @aerorahul this is the restart_interval spot, like you said, if we do bkg times in __init__, then this is not needed

File: ush/python/pygfs/task/aero_analysis.py (hunk 3 above)
Comment (aerorahul) [reply ush/python/pygfs/task/aero_analysis.py]: If you could move it to `__init__` in this PR itself, that would be great.

File: ush/python/pygfs/task/aero_analysis.py (hunk 5 above)
Comment (aerorahul) [reply ush/python/pygfs/task/aero_analysis.py]: `restart_interval` is a definition in `config.fcst` that controls the interval with which for the (atm) model to dump restarts.  (aerosols are output from the atm component).
It seems the use here is to identify what is the interval between backgrounds. The restarts happen to be the backgrounds for aeroDA.  Would using a different variable to disambiguate between the `restart_interval` that the model defines and the `background_interval` that the DA requires help?
Thoughts?

**Comment by WalterKolczynski-NOAA** [comment]
@CoryMartin-NOAA Any update on this?

**Comment by CoryMartin-NOAA** [comment]
@WalterKolczynski-NOAA let me see if I can make any progress in the next two days on the upgrades @aerorahul proposed.

**Comment by CoryMartin-NOAA** [comment]
@aerorahul did I make all the changes we talked about on Tuesday? I've tested and it works.

File: ush/python/pygfs/task/aero_analysis.py [hunk 6]
```diff
@@ -46,6 +46,7 @@ def __init__(self, config):
                 'npz_anl': self.config['LEVS'] - 1,
                 'AERO_WINDOW_BEGIN': _window_begin,
                 'AERO_WINDOW_LENGTH': f"PT{self.config['assim_freq']}H",
+                'aero_bkg_fhr': [int(t) for t in self.config['aero_bkg_times'].split(',')],
```
Comment (aerorahul) [inline ush/python/pygfs/task/aero_analysis.py]: `map` is a good way to do this without invoking `for` loops.
```suggestion
                'aero_bkg_fhr': map(int, self.config['aero_bkg_times'].split(',')),
```

**Review by aerorahul — APPROVED** [review]
Looks good.  
Just one suggestion to use `map` instead of `for` loop.

**Review by aerorahul — APPROVED** [review]

**Comment by CoryMartin-NOAA** [comment]
Good tip/suggestion, thanks @aerorahul for that and the merge!

**Pull request merged**